quesma/queryparser/pancake_json_rendering.go

```diff
@@ -0,0 +1,160 @@
+// Copyright Quesma, licensed under the Elastic License 2.0.
+// SPDX-License-Identifier: Elastic-2.0
+package queryparser
+
+import (
+	"context"
+	"fmt"
+	"quesma/model"
+	"quesma/util"
+	"strings"
+)
+
+type pancakeJSONRenderer struct {
+}
+
+func (p *pancakeJSONRenderer) selectMetricRows(metricName string, rows []model.QueryResultRow) (result []model.QueryResultRow) {
+	if len(rows) > 0 {
+		newRow := model.QueryResultRow{Index: rows[0].Index}
+		for _, col := range rows[0].Cols {
+			if strings.HasPrefix(col.ColName, metricName) {
+				newRow.Cols = append(newRow.Cols, col)
+			}
+		}
+		return []model.QueryResultRow{newRow}
+	}
+	return
+}
+
+func (p *pancakeJSONRenderer) splitBucketRows(bucketName string, rows []model.QueryResultRow) (
+	buckets []model.QueryResultRow, subAggrs [][]model.QueryResultRow) {
+
+	if len(rows) == 0 {
+		return buckets, subAggrs
+	}
+	bucketKeyName := bucketName + "key"
+	bucketCountName := bucketName + "count"
+	indexName := rows[0].Index
+	for rowIdx, row := range rows {
+		isNewBucket := rowIdx == 0 // first row is always new bucket
+		if !isNewBucket {          // for subsequent rows, create new bucket if any key is different
+			previousBucket := buckets[len(buckets)-1]
+			for _, cols := range row.Cols {
+				if strings.HasPrefix(cols.ColName, bucketKeyName) {
+					for _, previousCols := range previousBucket.Cols {
+						if cols.ColName == previousCols.ColName {
+							if cols.Value != previousCols.Value {
+								isNewBucket = true
+							}
+							break
+						}
+					}
+				}
+			}
+		}
+
+		if isNewBucket {
+			buckets = append(buckets, model.QueryResultRow{Index: indexName})
+			subAggrs = append(subAggrs, []model.QueryResultRow{})
+			lastIdx := len(buckets) - 1
+			for _, cols := range row.Cols {
+				if strings.HasPrefix(cols.ColName, bucketKeyName) || strings.HasPrefix(cols.ColName, bucketCountName) {
+					buckets[lastIdx].Cols = append(buckets[lastIdx].Cols, cols)
+				}
+			}
+		}
+		lastIdx := len(buckets) - 1
+		subAggrs[lastIdx] = append(subAggrs[lastIdx], row)
+	}
+
+	return buckets, subAggrs
+}
+
+// In some queries we want to filter out null values or empty
+// We accomplish that by increasing limit by one during SQL query and then filtering out during JSON rendering.
+// So we either filter out empty or last one if there is none.
+// This can't be replaced by WHERE in generic case.
+func (p *pancakeJSONRenderer) potentiallyRemoveExtraBucket(layer *pancakeAggregationLayer, bucketName string, bucketRows []model.QueryResultRow, subAggrRows [][]model.QueryResultRow) ([]model.QueryResultRow, [][]model.QueryResultRow) {
+	// We are filter out null
+	if layer.nextBucketAggregation.filterOurEmptyKeyBucket {
+		nullRowToDelete := -1
+		bucketKeyName := bucketName + "key"
+	ROW:
+		for i, row := range bucketRows {
+			for _, col := range row.Cols {
+				if strings.HasPrefix(col.ColName, bucketKeyName) {
```

Comment:
We can find the column index before processing rows and use it instead of looping over cols